components/cd-disclosure/cd-disclosure.js

```diff
@@ -0,0 +1,81 @@
+(function (Drupal) {
+  'use strict';
+
+  /**
+   * @see https://www.w3.org/WAI/ARIA/apg/patterns/disclosure/
+   */
+  Drupal.behaviors.cdDisclosure = {
+    idPrefix: 'cd-disclosure--',
+
+    attach: function (context, settings) {
+      // Initialize toggable dropdown.
+      this.initializeDisclosures();
+    },
+
+    /**
+     * Set up all Disclosure widgets found in the DOM.
+     *
+     * For our purposes, the `disclosure` element is the wrapped content itself
+     * and the button to toggle it is generated automatically by this script.
+     */
+    initializeDisclosures: function () {
+      let disclosures = document.querySelectorAll('.cd-disclosure');
+
+      // All behavior is controlled by data attributes.
+      disclosures.forEach((disclosure) => {
+        // Set a random ID if none is found.
+        let disclosureId = disclosure.getAttribute('id') || this.idPrefix + Math.floor(Math.random(1) * 100000000);
+
+        // Apply ID if it wasn't there already.
+        if (!disclosure.getAttribute('id')) {
+          disclosure.setAttribute('id', disclosureId);
+        }
+
+        // Determine where to place the button.
+        let positionAfter = disclosure.dataset.cdDisclosureInsert === 'after' || false;
+
+        // Determine whether content is shown by default. We consider this the
+        // edge-case so omitting the attribute assumes it is hidden.
+        let contentVisible = disclosure.dataset.cdDisclosureShowByDefault === 'true' || false;
+
+        // Extract button label
+        let buttonText = disclosure.dataset.cdDisclosureLabel || 'Toggle content';
```

Comment:
I can't remember where I read it now, but I found guidance that specifically advises against changing button labels according to state. The example I saw involved a play/pause button for a video player, where the label is "pause" when video plays, but "play" when it's paused.

This isn't the page, but it also avoids changing labels and instead offers a visual cue based on `aria-pressed`: https://www.w3.org/WAI/ARIA/apg/patterns/button/examples/button/

Additionally, the WAI Disclosure pattern also uses a visual cue instead of textual, in the form of a chevron. We could use our `aria-expanded` attribute to provide such a visual cue:
https://www.w3.org/WAI/ARIA/apg/patterns/disclosure/examples/disclosure-image-description/

At any rate, that's why I intentionally excluded the second label. But for now, to keep consistency with Dropdown I will add it in.

Reply:
**Changing labels?**

A recommendation about not changing the label is mentioned in https://inclusive-components.design/toggle-button/ ("Changing labels" section).

The reasoning is sound, yet that feels strange for the use on RW/UNO use case with the button after the content to reveal long content.

I couldn't find really a pattern for a show more/less type of disclosure with the button after the content.

We should probably rethink how it's handled on those sites. Not sure if we want to tackle that as part of the work on this component because we then need to involve the product owners if we change the design/behavior.

As proposed, one option is to add a parameter for a second label like in the cd-dropdown so that this new component can be used almost as is instead of the cd-dropdown on those sites.

However while this parameter would be optional, it's a potential open door to bad practice if changing label is considered as such.

**Alternative**

Alternatively, we could keep the single label attribute and add some extra javascript on those sites to listen to changes on the `aria-expanded` attribute for example to switch the labels to keep the current behavior. They already have a bit of complementary logic that could be extended for that.

This way this `cd-disclosure` component could follow the "best" practices and those sites could keep their current behavior until revisited.

@rupl What do you think?

Reply:
Nice find, yup that was probably the link I had in mind. I've read that page a couple times over the years.

Yes, we can go with the alternative you proposed. Out of curiosity, what do you have in mind to listen for the attribute's changed value? `MutationObserver`?

Reply:
Yes, like https://github.com/UN-OCHA/unocha-site/blob/develop/html/themes/custom/common_design_subtheme/templates/modules/unocha_donors/unocha-donors-list--oct-top-donors.html.twig.